kubernetes/cache.go

```diff
@@ -454,3 +670,118 @@ func (c *controllerImpl) GetServices(namespace string) ([]core_v1.Service, error
+			}
+			return osProjects, nil
+		}
+	}
+
+	return []osproject_v1.Project{}, nil
+}
+
+func (c *controllerImpl) GetNamespaces() ([]core_v1.Namespace, error) {
+	if err := c.checkStateAndRetry(); err != nil {
+		return []core_v1.Namespace{}, err
+	}
+
+	namespaces := c.projectInformer.GetIndexer().List()
+	if len(namespaces) > 0 {
+		if _, ok := namespaces[0].(*core_v1.Namespace); ok {
+			coreNamespaces := make([]core_v1.Namespace, len(namespaces))
+			for i, ns := range namespaces {
+				coreNamespaces[i] = *(ns.(*core_v1.Namespace))
+			}
+			return coreNamespaces, nil
+		}
+	}
+
+	return []core_v1.Namespace{}, nil
+
+}
+
+func createProjectsInformer(client *osproject_v1_client.ProjectV1Client, refreshDuration time.Duration) cache.SharedIndexInformer {
+	return cache.NewSharedIndexInformer(cache.NewListWatchFromClient(client.RESTClient(), "projects", meta_v1.NamespaceAll, fields.Everything()),
+		&osproject_v1.Project{},
+		refreshDuration,
+		cache.Indexers{},
+	)
+}
+
+func createNamespaceInformer(clientset kubernetes.Interface, refreshDuration time.Duration) cache.SharedIndexInformer {
+	sharedInformer := informers.NewSharedInformerFactory(clientset, refreshDuration)
+	return sharedInformer.Core().V1().Namespaces().Informer()
+}
+
+func (c *controllerImpl) GetVirtualServices(namespace string) (*GenericIstioObjectList, error) {
+	return c.getGenericIstioObjectList(namespace, virtualServiceType)
+}
+
+func (c *controllerImpl) getGenericIstioObjectList(namespace, istioType string) (*GenericIstioObjectList, error) {
+	// Do the parsing from GenericIstioObjectList to the correct one in istio_details_service.go so we don't replicate that code..
+	if err := c.checkStateAndRetry(); err != nil {
+		return nil, err
+	}
+	if _, found := c.istioControllers[namespace]; !found {
+		log.Debugf("No %s informers for %s found", istioType, namespace) // Debugf
+		return nil, fmt.Errorf("Istio caching is not enabled for namespace %s", namespace)
+	}
+
+	vss := c.istioControllers[namespace][istioType].GetIndexer().List()
+	log.Infof("Len of vss: %d for %s\n", len(vss), namespace)
+	if len(vss) > 0 {
+		log.Debugf("Received object of type: %s\n", reflect.TypeOf(vss[0]))
+		itemList := &GenericIstioObjectList{}
+		obj, ok := vss[0].(*GenericIstioObject)
+		if ok {
+			itemList.TypeMeta = obj.TypeMeta
+			objects := make([]GenericIstioObject, 0, len(vss))
+			for _, o := range vss {
+				obj, ok := o.(*GenericIstioObject)
+				if ok {
+					objects = append(objects, *obj)
+				}
+			}
+			itemList.Items = objects
+			log.Debugf("Returning items size: %d\n", len(itemList.Items))
+			return itemList, nil
+		}
+		log.Infof("Got %d %s items back from cache", len(vss), istioType) // Debugf
+	}
+
+	return nil, nil
+}
+
+// Allow subset of cache to be used - could be enabled all the way to namespace level
+// TODO There's no config option for namespace level configuration yet
+func (c *controllerImpl) KubeCached(namespace string) bool {
```

Comment:
Yeah, I guess if here also we may want to discriminate per type too, or not.
KubeCached("bookinfo", "VirtualServices")

I'm more sure that we can save some performance if we skip rare objects (i.e thinking in not used controllers for workloads and also not all Istio objects deserves to be cached as well).

Reply:
I don't know if it makes sense to have that distinction here. Unless we then allow dynamic configuration of types to be cached. If instead we just want to skip "rare objects", it's more simple to just not make any ``GetRareObject()`` methods in the cache.go

Reply:
How that affects things like fetching ``IstioConfigList`` is something we need to consider also. It still takes time to fetch those from the k8s API, even if they're not used for other things.